File: todo-api-rs/src/application/dtos/todo/list.rs
use std::error;
use std::num::NonZeroU32;

use thiserror::Error;

use crate::domain::entities::todo::{Title, TodoEntity};

#[derive(Clone, Debug)]
pub struct ListTodosInput {
    pub page: NonZeroU32,
    pub per_page: NonZeroU32,
    pub title: Option<Title>,
}

pub type ListTodosOutput = Result<TodosList, ListTodosError>;

#[derive(Clone, Debug)]
pub struct TodosList {
    pub count: u64,
    pub page: NonZeroU32,
    pub per_page: NonZeroU32,
    pub items: Vec<TodoEntity>,
}

#[derive(Debug, Error)]
pub enum ListTodosError {
    #[error(transparent)]
    Internal(Box<dyn error::Error>),
}

File: todo-api-js/src/application/dtos/todo/list.ts
import { type Result } from "@domain/utils/result";
import { type ListError } from "@application/repositories/todo";
import { type TodoEntity } from "@domain/entities/todo";

export type ListTodosInput = { page: number; limit: number };

export type ListTodosOutput = Result<TodoList, ListTodosError>;

export type TodoList = {
    count: number;
    page: number;
    limit: number;
    data: TodoEntity[];
};

export type ListTodosError = ListError;
